File: 16.3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    let closest = nums[0] + nums[1] + nums[2];
    nums.sort(function(a, b) {
        return a - b;
    });
    for (let i = 0; i < nums.length - 2; i++) {
        let j = i + 1, k = nums.length - 1;
        while (j < k) {
            let sum = nums[i] + nums[j] + nums[k];
            if (Math.abs(sum - target) < Math.abs(closest - target)) {
                closest = sum;
            }
            if (sum > target) {
                k--;
            } else if (sum < target) {
                j++;
            } else {
                return target;
            }
        }
    }
    return closest;
};

File: 16.3sum-closest.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        closest = nums[0] + nums[1] + nums[2];
        nums.sort()
        for i in range(len(nums) - 2):
            j, k = i + 1, len(nums) - 1
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if abs(sum - target) < abs(closest - target):
                    closest = sum
                if sum > target:
                    k -= 1
                elif sum < target:
                    j += 1
                else:
                    return target
        return closest